File: Leetcode75/Binary Search/2300. Successful Pairs of Spells and Potions/solution.cpp
class Solution {
public:
    vector<int> successfulPairs(vector<int>& spells, vector<int>& potions, long long success) {
        ios_base::sync_with_stdio(false); cout.tie(NULL); cin.tie(NULL);
        sort(potions.begin(), potions.end()); // Sort the potions vector
        vector<int> ans;
        for(int i = 0, sz = spells.size(); i < sz; i++) {
            int l = 0, r = potions.size() - 1; // Initialize left and right pointers for binary search
            while (l <= r) {
                int mid = l + (r - l) / 2; // Calculate the middle index
                // Check if the product of the current potion and spell is greater than or equal to success
                if ((long long)potions[mid] * spells[i] >= success) {
                    r = mid - 1; // If so, update the right pointer
                } else {
                    l = mid + 1; // Otherwise, update the left pointer
                }
            }
            // Push the count of successful pairs into the answer vector
            ans.push_back(potions.size() - l);
        }
        return ans;
    }
};

File: Leetcode75/Binary Search/2300. Successful Pairs of Spells and Potions/solution.py
class Solution:
    def successfulPairs(self, spells: List[int], potions: List[int], success: int) -> List[int]:
        
        # Sort the potions list in ascending order
        potions.sort()
        
        # Initialize an empty list to store the counts of successful pairs
        ans = list()
        
        # Get the length of the potions list
        ln = len(potions)
        
        # Iterate through each spell in the spells list
        for i in spells:
            # Initialize pointers for binary search
            l, r = 0, ln - 1 

            # Perform binary search to find the count of successful pairs for the current spell
            while l <= r:
                mid = (l + r) // 2
                
                # Check if the product of the current potion and the spell is greater than or equal to the success value
                if i * potions[mid] >= success:
                    r = mid - 1  # Update the right pointer to search in the left half
                else:
                    l = mid + 1  # Update the left pointer to search in the right half

            # Append the count of successful pairs for the current spell to the ans list
            ans.append(ln - l)
        
        # Return the list containing counts of successful pairs for each spell
        return ans
